src/snmalloc/ds_core/helpers.h

```diff
@@ -361,43 +339,33 @@ namespace snmalloc
       append(static_cast<unsigned long long>(x));
     }
 
-  public:
     /**
-     * Constructor.  Takes a format string and the arguments to output.
+     * Append with format string and arguments. Compiler
+     * is able to optimize the recursion into loops.
      */
-    template<typename... Args>
-    SNMALLOC_FAST_PATH MessageBuilder(const char* fmt, Args... args)
+    template<typename Head, typename... Tail>
+    void append(const char* fmt, Head&& head, Tail&&... tail)
     {
-      buffer[SafeLength] = 0;
-      size_t arg = 0;
-      auto args_tuple = std::forward_as_tuple(args...);
-      for (const char* s = fmt; *s != 0; ++s)
+      if (fmt[0] == '{' && fmt[1] == '}')
```

Comment:
I think we still need the case of

```C++
if (fmt[0] == 0)
  error("Internal error: format string missing {}!");
```

Otherwise the code will OOB read if the {} is missing.